### PR TITLE
Feature/sweep/support array paramtype

### PR DESCRIPTION
**Pull request by astafan8**
Allows specifying `paramtype` that is relevant for the way the values are saved. `paramtype` is a part of `ParamSpec` that controls how values are saved in sqlite database backend of qcodes.

This flexibility allows to save values as `array` in order to increase the speed of data saving.

The code is intentionally not pretty because a re-thinking of the framework is planned.

@sohailc would you like to have a look? :)
@wpfff

### CODE REVIEW

File: pytopo/sweep/decorators.py [hunk 1]
```diff
@@ -167,11 +200,17 @@ def wrapper() ->dict:
     return decorator
 
 
-def parameter_setter(parameter):
-    names_units = (parameter.full_name, parameter.unit)
+def parameter_setter(parameter, paramtype: str = None):
+    if paramtype:
+        names_units = (parameter.full_name, parameter.unit, paramtype)
+    else:
+        names_units = (parameter.full_name, parameter.unit)
```
Comment (astafan8) [inline pytopo/sweep/decorators.py]: perhaps, move these four lines into a separate function?

File: pytopo/sweep/decorators.py [hunk 2]
```diff
@@ -96,9 +129,9 @@ def hardsweep(ind: List[Tuple], dep: List[Tuple]) ->Callable:
...
     Args:
         ind: List of independent parameters, defined as tuples of names and
-                units.
+                units (and optionally 'paramtype').
         dep: List of dependent parameters, defined as tuples of names and
-                units.
+                units (and optionally 'paramtype').
 
     Returns:
         A decorator which returns a sweep object, which can be directly used
```
Comment (wpfff) [inline pytopo/sweep/decorators.py]: at least for the `hardsweep` decorator, the dictionaries that the resulting sweep object generates don't have numpy arrays. so this does not completely work as expected.

**Comment by astafan8** [comment]
Extra changes:
* `hardsweep` decorator works now for `'array'` parameters.
* `hardsweep` decorator is covered with tests that showcase the cases where it works and where it does not. In the future, this will be explained in the documentation. Of course, whenever necessary this will definitely be improved.
* `setter` and `getter` are also now covered with tests which also showcase their support for `'array'` parameters.

The code is a bit dirty and lacks proper documentation. Nevertheless, the current behavior of the framework is captured with tests, hence refactoring and improvements can be now done in a safer and more structured way. But these will be topics of other PRs.

@wpfff Could you try your code again after this update?

File: pytopo/sweep/decorators.py (hunk 2 above)
Comment (astafan8) [reply pytopo/sweep/decorators.py]: should be fixed after some commits.

**Pull request merged**